File: RayTracerAvalonia/RayTracing/Renderer.cs
using System;
using System.Threading.Tasks;

namespace RayTracerAvalonia.RayTracing;
public class Renderer(int CanvasWidth, int CanvasHeight)
{
    public unsafe void Render(Memory<byte> canvas, Scene scene)
    {
        var tempCanvas = canvas;
        var tempScene = scene;
        var handle = (byte*)canvas.Pin().Pointer;
        

        Parallel.For(0, CanvasHeight, (int h) =>
        {
            var y = ((float)h / CanvasHeight) - .5f;
            Parallel.For(0, CanvasWidth, (int w) =>
            {
                var x = ((float)w / CanvasWidth) - .5f;
                var color = tempScene.Trace(x, y);

                var index = (h * CanvasWidth + w) * 4;
                
                
                *(handle + index * sizeof(byte)) = color.B;
                *(handle + (index+1) * sizeof(byte))  = color.G;
                *(handle + (index+2) * sizeof(byte))  = color.R;
                *(handle + (index+3) * sizeof(byte))  = 255;
            });
        });
    }
}

File: RayTracerAvalonia.Desktop/cpp_raytracing_engine/renderer.cpp
#include "include/renderer.hpp"
#include "include/scene.hpp"
#include <algorithm>
#include <cstdint>
#include <print>

// Constructor implementation
Renderer::Renderer(uint16_t width, uint16_t height)
    : canvasHeight(height), canvasWidth(width) {}

// Render method implementation
void Renderer::Render(uint8_t *data, const Scene *scene) const noexcept {
  // Parallelize using OpenMP if available (C++23 std::execution can be used for
  // more advanced parallelism)

#pragma omp parallel for
  for (uint16_t h = 0; h < canvasHeight; ++h) {

    for (uint16_t w = 0; w < canvasWidth; ++w) {
      auto x = (static_cast<float>(w) / canvasWidth) - 0.5f;
      auto y = (static_cast<float>(h) / canvasHeight) - 0.5f;

      auto color = scene->Trace(x, y);
      auto mw = static_cast<uint16_t>(canvasWidth - 1 - w);

      auto index = (h * canvasWidth + mw) * 4;

      data[index] = color.b;
      data[index + 1] = color.g;
      data[index + 2] = color.r;
      data[index + 3] = color.a;
    }
  }
}
